### PR TITLE
feat/rmrk2.0

### PR DESCRIPTION
**Pull request by KngZhi**
this PR is related with #42 

current progress
- [x] Implement interactions only v2 has.
- [x] Check Existing Interactions
- [x] backward compatible with version1.0's interaction
- [x] Ability to create this v2 interactions
  - [x] create Interaction Event
  - [x] createCollection, createNFT
  - [x] create Base
- [ ] tsdoc for provide example.

### CODE REVIEW

**Comment by KngZhi** [comment]
@vikiival mind make a review for submitted code, to make sure things going on the right path and code style :)

Also made very specific type for every interaction, since v2 standard has more than `value` field, and wants to import on `Rubick` later

**Comment by vikiival** [comment]
Mark as unread, will check tomorrow!

**Comment by KngZhi** [comment]
> Mark as unread, will check tomorrow!

😉

**Comment by KngZhi** [comment]
hey, @vikiival mind making a review?

**Comment by vikiival** [comment]
> hey, @vikiival mind making a review?

Yup on it :)

File: minimark/src/rmrk/helpers.ts [hunk 1]
```diff
@@ -9,7 +9,8 @@ export const isRemark = (text: string): boolean => {
 export const splitBySquare = (text: string): string[] => text.split(SQUARE)
 
 export const isValidInteraction = (interaction: string, throwable = true): boolean => {
-  const value = (Object.values(Interaction) as string[]).includes(interaction)
+  const availableInteractions: string[] = [...Object.values(Interaction), ...Object.values(InteractionV2)]
```
Comment (vikiival) [inline minimark/src/rmrk/helpers.ts:12]: Merge the object into constants

File: minimark/src/rmrk/types.ts [hunk 2]
```diff
@@ -11,20 +11,70 @@ export enum Interaction {
...
+  replace: string
+}
+
+export type Setpriority = Required<InteractionValue>
+
+export type Setproperty = Required<InteractionValue> & {
+  name: string
+}
+
+export type Themeadd = Required<InteractionValue> & {
```
Comment (vikiival) [inline minimark/src/rmrk/types.ts:72]:
```suggestion
export type AddTheme = Required<InteractionValue> & {
```

File: minimark/src/rmrk/types.ts [hunk 3]
```diff
@@ -11,20 +11,70 @@ export enum Interaction {
...
+}
+
+export type Lock = Omit<InteractionValue, 'value'>
+
+export type Resadd = Required<InteractionValue> & {
+  value: Record<string, string>
+  replace: string
+}
+
+export type Setpriority = Required<InteractionValue>
```
Comment (vikiival) [inline minimark/src/rmrk/types.ts:66]:
```suggestion
export type SetPriority = Required<InteractionValue>
```

File: minimark/src/rmrk/types.ts [hunk 4]
```diff
@@ -11,20 +11,70 @@ export enum Interaction {
...
+
+export type Setproperty = Required<InteractionValue> & {
```
Comment (vikiival) [inline minimark/src/rmrk/types.ts:68]:
```suggestion
export type SetProperty = Required<InteractionValue> & {
```

**Review by vikiival — CHANGES_REQUESTED** [review]
Nice draft!

Please move all the logic and types to folder `src/rmrk/v2`

**Comment by vikiival** [comment]
+ Not sure if tests are correct 👀

File: minimark/src/rmrk/types.ts (hunk 2 above)
Comment (KngZhi) [reply minimark/src/rmrk/types.ts]: This is what the standard is named, in order to consistency, I would revert that to `ThemeAdd`[1]

https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/interactions/themeadd.md

**Comment by KngZhi** [comment]
Some principles are followed here,

1. avoid `any` in any exposed API, meanwhile provide specific types for the user
2. break early. Throw `Error` for wrong params, since if that error goes on-chain, could be hard to debug or uninformative debug info.
3. any parameter or property should exactly the same as RMRK2.0.0 standards.

**Comment by yangwao** [comment]
@vikiival let's review this 👀

**Comment by vikiival** [comment]
@KngZhi Sir is this possible to namespace it ? 

so consumer of this package can do something like:
```js
import { createInteraction } from 'minimark/v2'
```

**Comment by vikiival** [comment]
It's really disturbing to have import under one

File: minimark/src/rmrk/v2/create.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    case InteractionV2.THEMEADD:
+      return convert([payload.base_id, payload.name, wrapToString(payload.value)])
+    default:
+      throw new Error(`Unsupported action: ${action}`);
+  }
+
+  throw new Error(`Unsupported action: ${action}`);
+}
+
+type CreateNFTProps = {
```
Comment (vikiival) [inline minimark/src/rmrk/v2/create.ts:58]: Please do not mix types and code

**Comment by KngZhi** [comment]
> @KngZhi Sir is this possible to namespace it ?
> 
> so consumer of this package can do something like:
> 
> ```js
> import { createInteraction } from 'minimark/v2'
> ```

Yes sir, would be same with v1 version.

File: minimark/src/rmrk/v2/types.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+export type LIST = {
+    id: string,
+    price: string,
+}
+
+export type CHANGEISSUER = {
+    id: string;
+    newissuer: string;
```
Comment (vikiival) [inline minimark/src/rmrk/v2/types.ts:77]: Cammel Case

**Review by vikiival — CHANGES_REQUESTED** [review]
👀 

Please make sure that you coding style and variable naming is consistent

File: minimark/src/rmrk/v2/create.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    symbol,
+    max,
+    metadata,
+  } = props || {}
+
+  if (max < 0) {
+    throw new Error("max should be equal or greater than zero");
+  }
+  const theSymbol = makeSymbol(symbol)
+  return {
```
[the remark this answers is not in the document]
Comment (KngZhi) [reply minimark/src/rmrk/v2/create.ts:105]: Name goes to the `metadata` field. [[ref]]

example of collection field
```json
{
  "max": 100,
  "issuer": "CpjsLDC1JFyrhm3ftC9Gs4QoyrkHKhZKtK7YqGTRFtTafgp",
  "symbol": "DLEP",
  "id": "0aff6865bed3a66b-DLEP",
  "metadata": "ipfs://ipfs/QmVgs8P4awhZpFXhkkgnCwBp4AdKRj3F9K58mCZ6fxvn3j"
}
```


[ref]: https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/entities/collection.md#metadata

**Comment by vikiival** [comment]
> align the names of parameters with the RMRK2.0.0

I am fine with that :) 

> The reason for using props way is these create functions have more than 3 parameters, dev may not know which parameters to what

Understood. but still creating string make it not friendly

**Comment by KngZhi** [comment]
> > align the names of parameters with the RMRK2.0.0
> 
> I am fine with that :)
> 
> > The reason for using props way is these create functions have more than 3 parameters, dev may not know which parameters to what
> 
> Understood. but still creating string make it not friendly

Thanks. So we are good to go? :)

**Comment by yangwao** [comment]
@vikiival is watching RMRK2.0 to completely understand stuff, so we don't miss some issues having it :)
I guess will be back on this shortly

**Comment by joelamouche** [comment]
@vikiival Okay please give me some instructions on how to run this locally and I will test it

**Comment by vikiival** [comment]
> @vikiival Okay please give me some instructions on how to run this locally and I will test it

1. Clone this repo
2. `gh pr checkout 43`
3. `cd minimark`
4. `pnpm i`
5. `pnpm build`

I can publish suite to that would behave as express server. So we can test if it creates correct data.

+ @joelamouche is it possible to get somehow `remarks` that were stored on-chain for substraknights? It would help us with unit tests since you know the best what should be the correct output

**Comment by vikiival** [comment]
rest of the work will continue in branch `remarkable-charon`

**Comment by vikiival** [comment]
#91

**Pull request merged**